rfcs/012-fdb-reduce.md

```diff
@@ -0,0 +1,190 @@
+                    emit([doc.year, doc.month, doc.day], 1);
+                },
+
+                "reduce": "_count"
+            }
+        }
+    }
+
+```
+
+And it emits the following key/value results for the reduce function:
+
+```js
+    [2017, 03, 1] = 1
+    [2017, 04, 1] = 1
+    [2017, 04, 15] = 1
+    [2017, 05, 1] = 1
+
+    [2018, 03, 1] = 1
+    [2018, 04, 1] = 1
+    [2018, 05, 1] = 1
+
+    [2019, 03, 1] = 1
+    [2018, 04, 1] = 1
+    [2018, 05, 1] = 1
+```
+
+#### Create
+
+To build the skip list, all keys will be added to level 0. Then each level up, a reduced number of keys will be added. For each level above 0, if a key/value is not added to that level, then that key's value is aggregated with the previous node in that row. Therefore each key/value node in a level is an aggregation of its key/value in level 0 and any key/values from the previous level that are greater than the node and smaller than the next node on that level. 
+
+See figure 2 for an example of the listed keys added to a skip list.
+
+![figure 2: The emitted reduce keys added to a reduce skip list](images/SkExample2.png)
+*figure 2:* The emitted reduce keys added to a reduce skip lis
+
+The optimal number of levels and the correct distribution algorithm is unknown at this stage, these settings will be configurable so that we can test and determine the best fit.
+
+#### Query
+
+From the above figure 2, we can see that for a reduce query with `group = true` then level 0 will be used to return all the exact keys.
+And for a query with `group_level = 0` then the highest level can be used. 
+If a `group_level > 1` is set for a reduce query, we need to traverse the skip list and aggregate the results before returning the results to the user.
+
+For example, using figure 2, with a reduce query of `group_level = 2`. We start at Level 4, we would traverse down to level 3, compare Node 0 and Node [2018, 03, 1]. They are not the same key for `group_level = 2` so we need to move to a lower level. We would perform the same action of comparing the current node with the next node to see if they are the same key until we find matching nodes or we get to level 0. In this case, we would reach level 0 and return Node [2017, 03, 1]. At Node [2017, 04, 1] it is possible to go back up to level 2 where we can compare [2017, 04, 1] and [2017, 04, 15]. Given they are the same, we can continue on level 1 and collect all the [2017, 04, x] keys. Once we have collected all of those. Those keys are run through a re-reduce before being returned. This process of comparing the current node with the next node would continue onwards. Each time comparing the current node with the next node. If they are the same, compare the Nodes at the next Level above until we reach a level where the keys are not equal. At that point, we would use the node values at the level just below that to move across, collecting all the equal keys to be run through a reducer before being returned.
```

Comment:
Out of curiosity, why is the first Node `0` instead of the first key?

Reply:
I based that off of the rankedset implementation's I've seen. I'm definitely open to rather making it the first key. At this stage, I'm not sure which would be better.

Reply:
Coming back to this. Having the first Node as `0` makes the starting point quite predictable and when we querying from the beginning of the index, one less key fetch to determine where to start.

Reply:
Just adding another point on this because I was testing the implementation without having a starting point. Another issue we get here if we don't specify a starting every time we get a key that is the lowest in the index we would have to add it to all skiplevels. This means we would get a clump of keys at the beginning of the index.